stan/math/prim/mat/fun/append_array.hpp

```diff
@@ -3,78 +3,216 @@
 
 #include <stan/math/prim/mat/fun/Eigen.hpp>
 #include <stan/math/prim/scal/meta/return_type.hpp>
+#include <stan/math/prim/scal/err/check_consistent_sizes.hpp>
 #include <vector>
+#include <iostream>
 
 namespace stan {
   namespace math {
+    /**
+     * This is custom return type promotion logic for append_array.
+     *
+     * By default, the type member is computed from the return_type metaprogram.
+     *
+     * Specializations are provided for std::vector and Eigen::Matrix so that
+     * the promotion happens only on the scalar types inside those containers
+     * without changing the containers themselves. Roughly,
+     * std::vector<T1>, std::vector<T2> promotes to
+     * std::vector<return_type<T1, T2>>.
+     *
+     * @tparam T1 First type to be promoted
+     * @tparam T2 Second type to be promoted
+     */
+    template <typename T1, typename T2>
+    struct append_return_type {
+      typedef typename return_type<T1, T2>::type type;
+    };
 
     /**
-     * Return the concatenation of two specified integer vectors in the
-     *   order of the arguments.
+     * Part of return type promotion logic for append_array
      *
-     * @param x First vector
-     * @param y Second vector
-     * @return A vector of x and y concatenated together (in that order)
+     * return_type<int, int> promotes to double, which is not the desired
+     * behavior (hence this specialization)
+     *
+     * @tparam T1 First type to be promoted
+     * @tparam T2 Second type to be promoted
      */
-    inline std::vector<int>
-    append_array(const std::vector<int>& x, const std::vector<int>& y) {
-      std::vector<int> z;
-      z.reserve(x.size() + y.size());
-      z.insert(z.end(), x.begin(), x.end());
-      z.insert(z.end(), y.begin(), y.end());
-      return z;
-    }
+    template <>
+    struct append_return_type<int, int> {
```

Comment:
maybe just a bit of doc here or in the last general version to indicate that this version is needed because `return_type<int, int>::type` is `double` (or at least I think that's what return type does).